lib/connect/api.js

```diff
@@ -1,89 +1,73 @@
-  let withBody = false
-  let headers = {
-    'Content-Type': 'application/json',
-    'Authorization': 'Bearer ' + context.auth.password,
-    'Heroku-Client': 'cli'
+let request = exports.request = function (context, method, url, data) {
+  if (!context.region) {
+    throw new Error('Must provide region URL')
   }
+  const connectClient = makeConnectClient(context)
+  return connectClient({
+    baseURL: context.region,
+    method,
+    url,
+    data
+  })
+}
 
-  // add data as querystring for GET request
-  if (data !== undefined && method === 'GET') {
-    path += ('?' + querystring.stringify(data))
-  }
-  // add data to body for POST/PUT requests
-  if (data !== undefined && ['POST', 'PUT', 'PATCH'].indexOf(method) !== -1) {
-    withBody = true
-    data = JSON.stringify(data)
-    headers['Content-Length'] = data.length
-  }
-
-  let hostname = regions.urlFromRegion(context.region)
-
-  let promise = new Promise(function (resolve, reject) {
-    let req = https.request({
-      hostname: hostname,
-      path: path,
-      method: method,
-      headers: headers
-    }, function (res) {
-      let body = ''
+let withUserConnections = exports.withUserConnections = co.wrap(function * (context, appName, flags, allowNone, heroku) {
+  const discoveryClient = makeDiscoveryClient(context)
+  const connectClient = makeConnectClient(context)
+  const resourceFlag = context.flags.resource
+  const resourceFlagQueryParam = resourceFlag ? `&resource_name=${resourceFlag}` : ''
 
-      res.setEncoding('utf8')
-      res.on('data', function (chunk) {
-        body += chunk
-      })
+  const response = yield discoveryClient(`/connections?app=${appName}${resourceFlagQueryParam}`)
+  const connections = response.data.results
 
-      res.on('end', function () {
-        // can only resolve with one argument, add json data to response object
-        try {
-          res.json = JSON.parse(body)
-        } catch (e) {
-          // body is empty, like with a DELETE, or invalid json
-          res.json = null
-          res.text = body
-        }
-        if (res.statusCode >= 400) {
-          let message = res.json ? res.json['message'] : res.statusCode + ' ' + res.statusMessage + ': ' + res.text
-          reject(new Error(message))
-        } else {
-          resolve(res)
-        }
-      })
-    })
+  if (connections.length === 0) {
+    return yield Promise.resolve([])
+  }
 
-    req.on('error', function (err) {
-      console.log('Request error', err)
-      reject(err)
-    })
+  const fetchConnectionDetailFuncs = connections.map(c => connectClient.get(`${c.detail_url}?deep=true`))
 
-    // add data to body for POST/PUT requests
-    if (withBody) {
-      req.write(data)
+  const aggregatedConnectionsResponse = yield Promise.all(fetchConnectionDetailFuncs)
+  const aggregatedConnectionsData = aggregatedConnectionsResponse.map(resp => {
+    return {
+      ...resp.data,
+      ...connections.find(c => c.id === resp.data.id)
```

Comment:
Things like this remind me just how long it's been since I've written JavaScript regularly. Very cool!

Reply:
Though now that I look in more detail, it looks like withConnection still has to fetch all the connections in order to narrow down just the one it needs. Is that intentional? Especially now that fetching all the connection info is an N+1 operation?